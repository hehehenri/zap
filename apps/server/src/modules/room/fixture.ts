import { RoomDefinition, RoomModel } from "./RoomModel";

type Fields = Pick<RoomDefinition, Exclude<keyof RoomDefinition, "_id">>

export const createRoom = async (fields?: DeepPartial<Fields>) => {
  return await new RoomModel(fields).save();
}
