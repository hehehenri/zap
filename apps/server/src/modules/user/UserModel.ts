import mongoose from "mongoose";

export type UserDefinition = {
  readonly _id: mongoose.Types.ObjectId;
  readonly username: string
  readonly password: string
}

export const userSchema = new mongoose.Schema<UserDefinition>(
  {
    username: { type: String, required: true },
    password: { type: String, required: true, select: false },
  },
  { collection: 'User' }
);

export const UserModel = mongoose.model<UserDefinition>('User', userSchema);
