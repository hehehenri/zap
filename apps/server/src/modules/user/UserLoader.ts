import { createLoader } from "@entria/graphql-mongo-helpers";
import { UserModel } from "./UserModel";

export const UserLoader = createLoader({ model: UserModel, loaderName: "UserLoader" });
