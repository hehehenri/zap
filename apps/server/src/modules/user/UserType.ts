import { GraphQLNonNull, GraphQLObjectType, GraphQLString } from "graphql";
import { UserDefinition } from "./UserModel";
import { connectionDefinitions } from "graphql-relay";

const UserType = new GraphQLObjectType<UserDefinition>({
  name: "User",
  fields: () => ({
    id: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (user) => user._id.toString()
    },
    username: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (user) => user.username
    }
  })
});

export const UserConnection = connectionDefinitions({
  name: "User",
  nodeType: UserType,
});

export default UserType;
