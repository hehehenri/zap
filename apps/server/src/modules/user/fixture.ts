import config from "../../config";
import { UserDefinition, UserModel } from "./UserModel"
import bcrypt from "bcryptjs";

let counter = 0;

const getCounter = () => {
  counter += 1;

  return counter;
}

type Fields = Pick<UserDefinition, Exclude<keyof UserDefinition, "_id">>

export const createUser = async (fields?: DeepPartial<Fields>) => {
  const index = getCounter();

  const plainTextPassword = fields?.password ?? "password";

  const password = bcrypt.hashSync(plainTextPassword, config.jwt.saltRounds);
  const username = fields?.username ?? `user#${index}`;

  const user = new UserModel({ password, username });
  await user.save();

  return user;
}
