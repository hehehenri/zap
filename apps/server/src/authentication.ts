import jwt from "jsonwebtoken";
import { UserModel, UserDocument } from "./modules/user/UserModel";
import config from "./config";
import { KoaContext } from "./context";

export type Config = {
  secret: string
}

type TokenPayload = {
  id: string
}

export const getToken = (ctx: KoaContext) => {
  const auth = ctx.headers.authorization;
  if (!auth) return null;

  if (auth.startsWith("Bearer ")) {
    return auth.substring(7, auth.length);
  }

  return null
}

export const getAuth = async (token: string | null) => {
  if (!token) return { user: null };

  try {
    const decoded = jwt.verify(token, config.jwt.secret) as TokenPayload;

    const user = await UserModel.findById(decoded.id);
    return { user };
  } catch (_) {
    return { user: null };
  }
}

export const generateToken = (user: UserDocument) => {
  return jwt.sign({ id: user._id }, config.jwt.secret);
}
