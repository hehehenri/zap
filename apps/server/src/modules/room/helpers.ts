import { UserDocument } from "../user/UserModel";
import { RoomDocument } from "./RoomModel";

export const isRoomMember = (room: RoomDocument, user: UserDocument) => {
  return room
    .participants
    .map(u => u._id.toString())
    .includes(user._id.toString());
}
