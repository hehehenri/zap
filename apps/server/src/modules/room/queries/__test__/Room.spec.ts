
import { toGlobalId } from "graphql-relay";
import { generateToken } from "../../../../authentication";
import { describeWithDb, testQuery } from "../../../../test/helpers";
import { createUser } from "../../../user/fixture";
import { createRoom } from "../../fixture";

const query = (roomId: string) => ({
  query: `
    query Room($roomId: ID!) {
      room(roomId: $roomId) {
        participants { username }
      }
    }
  `,
  variables: { roomId }
})

describeWithDb("room/queries/room", () => {
  it("should return room", async () => {
    const userA = await createUser({ username: "roomusera" });
    const userB = await createUser({ username: "roomuserb" });

    const room = await createRoom({ participants: [userA, userB] });

    const token = generateToken(userA);

    const response = await testQuery({ query: query(toGlobalId("Room", room._id)), token });

    expect(response.status).toBe(200);
    expect(response.body.data).toMatchSnapshot();
  });

  it("should not return room if user is not a member", async () => {
    const userA = await createUser({ username: "roomsusera" });
    const userB = await createUser({ username: "roomsuserb" });
    const userC = await createUser({ username: "roomsuserc" });

    const roomBC = await createRoom({ participants: [userB, userC] });

    const token = generateToken(userA);

    const response = await testQuery({ query: query(toGlobalId("Room", roomBC._id)), token });

    expect(response.status).toBe(422);
  });
});
