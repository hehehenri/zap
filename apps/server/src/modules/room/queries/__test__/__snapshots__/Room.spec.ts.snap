// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`room/queries/room should return room 1`] = `
{
  "room": {
    "participants": [
      {
        "username": "roomusera",
      },
      {
        "username": "roomuserb",
      },
    ],
  },
}
`;
